aaelf64-morello/aaelf64-morello.rst

```diff
@@ -827,3 +844,91 @@ veneer is used. The BX changes the execution state from A64 to C64:
+
+TLS for the pure capability ABI
+-------------------------------
+
+The design is based on TLSDESC, with the purpose of minimizing the performance
+differences between A64 and C64, while providing strict bounds when resolving
+TLS globals. Only the General Dynamic access model and the relaxation from
+General Dynamic to Local Exec are currently specified. However, this leaves the
+opportunity to add support for the other access models in the future.
+
+TLS static block
+^^^^^^^^^^^^^^^^
+
+The static block layout is the same used in AArch64 (Variant 1, see [TLS_]), with
+the only exception that TCB and the DTV pointer are capabilities.
+
+Thread pointer
+^^^^^^^^^^^^^^
+
+The thread pointer is a capability, held in ``CTPIDR_EL0``. The thread pointer
+needs to have the read, write, read capability and write capability permissions
+and bounds such that the entire TLS static block is accessible.
+
+Resolver functions
+^^^^^^^^^^^^^^^^^^
+
+A resolver function takes arguments in c0 (address of the TLS GOT slot), and c2
+(a copy of the thread pointer) and returns a pointer to the TLS global in c0
+and its size in x1. The resolver function has a custom calling convention that
+must preserve all registers except c0 and c1.
+
+Considerations:
+
+- Any dynamically loaded modules will be placed outside of the bounds of the
+  thread pointer, so a resolver function cannot return an offset from the
+  thread pointer, but rather needs to return a pointer (capability).
+- To minimize reading of ``CTPIDR_EL0``, the resolver functions take a
+  copy of ``CTPIDR_EL0`` as an argument and preserve it.
+- Not setting the bounds directly in the resolver allows shorter sequences
+  for the General Dynamic access model.
+
+
+Static TLS block resolver
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If the TLS variable is in the static block, while resolving the
+``R_MORELLO_TLSDESC`` relocation, the dynamic linker will place in the two GOT
+slots associated with this variable:
+
+- A capability to the static TLS block resolver function at offset 0.
+- The offset of the variable in the static TLS block at offset 16 (8 bytes).
+- The size of the variable at offset 24 (8 bytes).
+
+An implementation of the static block resolver could be the following:
+
+.. code-block:: text
+
+  ldp x0, x1 [c0, #16]
+  add c0, c2, x0
+  ret c30
+
+General Dynamic
+^^^^^^^^^^^^^^^
+
+The instruction sequence used for the General Dynamic access model is similar to
+that of other TLSDESC implementations, with the exception that the result
+doesn't need to be added to the thread pointer. However c2 needs to contain
+the thread pointer. The instruction sequence contains an additional NOP
+instruction in order to permit the static linker to relax to Local Exec.
+
+.. code-block:: text
+
+  adrp c0, :tlsdesc:sym
+  ldr c1, [c0, :tlsdesc_lo12:sym]
+  add c0, c0, :tlsdesc_lo12:sym
+  .tlsdesccall sym
+  blr c1
+  nop
+
+The compiler will emit a bounds setting instruction on c0 using the size
+returned in x1.
```

Comment:
How does that work with weak undef symbols? Are you relying on the fact that SCBNDS doesn't trap in Morello? If so that's really ugly, and another reason other than "the runtime should set bounds in the resolver not rely on the caller being correct" (otherwise you can just manually call a resolver from assembly and get back an overly-permissive capability) why this is not a good idea.

Reply:
Undef weak symbols should be handled using a special resolver that writes 0 to x0 and x1 (the text needs to updated).
This is indeed specific to Morello (scbnds won't trap).
If we would set the bounds in the resolver I guess the caller could still get the module base symbol if they wanted to get an over-permissive capability. I agree that setting the bounds in the resolver would reduce the permissions of the capabilities that are in registers (for the general case), so this needs further consideration.